### ISSUE TITLE
Add support to change the log level programmatically at runtime

### ISSUE THREAD

**Issue by daneshk** [ballerina-platform/ballerina-standard-library#4633]
**Description:**
We need a new API to change the log level programmatically at runtime without restarting the application. At the moment, the only way of changing the log level is from the `Config.toml` file that needs to restart the application. We need a new API like, `log:setLevel(log:DEBUG)` to change the level programmatically.

**Comment by ImeshBalasuriya** [comment]
Hi, 

I'd like to contribute to this issue. :)

**Comment by ImeshBalasuriya** [comment]
Hi @daneshk,

I have linked a PR that adds the required functionality. I am unsure of how to add integration tests that cover the new feature and I would gladly accept any pointers on how to do so.

**Comment by ImeshBalasuriya** [comment]
Hi,

Any update regarding this?

**Comment by daneshk** [comment]
Thanks, @ImeshBalasuriya for the contribution, and sorry for the delay in responding. I will go through your fix and get back to you soon.

**Comment by daneshk** [comment]
> Hi @daneshk,
> 
> I have linked a PR that adds the required functionality. I am unsure of how to add integration tests that cover the new feature and I would gladly accept any pointers on how to do so.

You can write an integration test like https://github.com/ballerina-platform/module-ballerina-log/blob/master/integration-tests/tests/tests_logfmt.bal#L94 

Here we run sample Ballerina projects and check the console output.

**Comment by keizer619** [comment]
@ImeshBalasuriya Are you working on this?

**Comment by ImeshBalasuriya** [comment]
Hi @keizer619,

Unfortunately, I'm a bit occupied atm. If this feature is urgent, please feel free to reassign it. If not, I am happy to work on it in December once I am free. 🙃

**Comment by daneshk** [comment]
This PR is merged in a different branch for the moment. We will keep this issue open till we write test cases for this feature.

**Comment by ovindu-a** [comment]
Is there any work to be done on this issue? @keizer619 @daneshk , If so I would like to work on it.

**Comment by Atharva1723** [comment]
Can I work on this issue?

**Comment by ovindu-a** [comment]
@daneshk shall I write the test cases for this issue? Or is there anything else to do?

**Comment by daneshk** [comment]
Hi @ovindu-a, Sorry for the delay in responding to your comment. Yes, we need to add test cases for this feature. Additionally, I would like to suggest a slight change in the log API as follows.

```ballerina
public isolated function setLogLevel(string logLevel, *ModuleName moduleName);

type ModuleName record {
    string organization;
    string module;
};
```

This will validate whether both the organization and module parts are provided correctly for the module name at compile time. It will allow users to pass the `moduleName` in the following ways.

```ballerina
    setLogLevel("DEBUG", organization = "ws", module = "test");
    setLogLevel("DEBUG", {organization: "ws", module: "test2"});
    setLogLevel("DEBUG", moduleName = {organization: "ws", module: "test3"});
```

Can you make this change in the API and add test cases to cover all scenarios? Please let me know if you have any questions.

**Comment by ovindu-a** [comment]
Sure @daneshk I will do the needful

**Comment by daneshk** [comment]
@ovindu-a There is an issue with the proposed API change, as we always need to set moduleName, which should not be the case at the global log level. So please make the changes as follows,

```ballerina
public isolated function setLogLevel(string logLevel, ModuleName? moduleName = ());

type ModuleName record {|
    string organization;
    string module;
|};
```

**Comment by ovindu-a** [comment]
@daneshk I have changed the function as requested. However since moduleName is made defaultable, I am not allowed to make it as an included record parameter. So the below code would not work right?

```
    setLogLevel("DEBUG", organization = "ws", module = "test");
```

Is this functionality okay. If so I will make the test cases.

**Comment by daneshk** [comment]
Yes, correct, @ovindu-a. Since the moduleName is an optional param, we have to make it defaultable rather than an included record parameter.

**Comment by ovindu-a** [comment]
@daneshk I'm currently writing the test cases, I have a small question

```ballerina 
@test:Config {}
public function testSetLogLevel() returns error? {
    Process|error execResultError = exec(bal_exec_path, {BAL_CONFIG_FILES: CONFIG_DEBUG_LOGFMT}, (), "run", SET_LOG_LEVEL_FILE);
    Process resultError = check execResultError;
    int _ = check resultError.waitForExit();
    int _ = check resultError.exitCode();

    io:ReadableByteChannel readableResultError = resultError.stderr();
    io:ReadableCharacterChannel scError = new (readableResultError, UTF_8);
    string outTextError = check scError.read(100000);
    string[] logLinesError = re`\n`.split(outTextError.trim());

    test:assertEquals(logLinesError.length(), 6, INCORRECT_NUMBER_OF_LINES);

    validateLog(logLinesError[5], MESSAGE_ERROR_LOGFMT);
}
```

This test will set the log level as ```DEBUG``` on in the config, but the test case expects only errors to be logged.

So if setting the log level using the function below works, the test will pass.

```ballerina
import ballerina/log;

public function main() {
    log:setLogLevel(ERROR);
    log:printError("error log");
    log:printWarn("warn log");
    log:printInfo("info log");
    log:printDebug("debug log");
}
```

Only issue is that I currently get an error when calling the ```setLogLevel``` function. But this should work when ```setLogLevel``` is included in the ballerina log library.

Is there a workaround to test this now or is it alright to keep it as it is.

**Comment by ovindu-a** [comment]
@daneshk I'm running the tests with ``` ./gradlew clean test -Pgroups=ballerinaIntegrationTests ```

I'm getting 2 errors in the line 
```ballerina
    log:setLogLevel(ERROR);
```

undefined function 'setLogLevel'(BCE2011)
undefined symbol 'ERROR'(BCE2010)

**Comment by daneshk** [comment]
Sorry @ovindu-a for late reply. 

The `integration-tests` is a separate module, in order to use the enum we need to use the module prefix(`log:`) like below,
```
log:setLogLevel(log:ERROR);
```
I am could not find the cause why the new API not picking up for the testcase yet. Can you run `./gradlew clean build -Pgroups=ballerinaIntegrationTests` which will build the latest code and test.

I would like to propose one small suggestion to the new API to use `log:LogLevel` as the first param instead of `string` type. So we can restrict the log levels in our new API. like below,

```ballerina
public isolated function setLogLevel(LogLevel logLevel, ModuleName? moduleName = ());
```

Can you also check this.

**Comment by ovindu-a** [comment]
Noted @daneshk, I will work on this

**Comment by daneshk** [comment]
Hi @ovindu-a, Could you able to write integration test cases?

**Comment by ovindu-a** [comment]
Currently working on it @daneshk. Hope to finish it by weekend.

**Comment by ovindu-a** [comment]
> I would like to propose one small suggestion to the new API to use `log:LogLevel` as the first param instead of `string` type. So we can restrict the log levels in our new API. like below,
> 
> ```ballerina
> public isolated function setLogLevel(LogLevel logLevel, ModuleName? moduleName = ());
> ```
> 
> Can you also check this.

The above change has been made @daneshk 

I have also written a few test cases, however I'm still having the issue of the APIs not being picked up for the testcase.

**Comment by daneshk** [comment]
@ovindu-a Can you send the PR with the failing test cases So I can check it?

**Comment by keizer619** [comment]
@ovindu-a We are opening this issue for public. If you are willing to work on this let us know

**Comment by Ctoic** [comment]
Hi @keizer619 I want to work on this issue please assign it to me
